File: Python/mesh.py
from scipy import sparse
import numpy as np

# %% Triangles

class Triangle:

    def __init__(self,nXY):
        
        self.M = 0
        self.Sx = 0
        self.Sy = 0
        self.nXY = nXY
        self.gRS = np.zeros((4,2))
        self.wei = np.array([-27,25,25,25])/96
        
        # Integration points coordinates
        
        self.gRS[0] = [1/3,1/3]
        self.gRS[1] = [0.6,0.2]
        self.gRS[2] = [0.2,0.6]
        self.gRS[3] = [0.2,0.2]
        
        # Computes the Jacobian matrix
        
        J11 = -self.nXY[0,0]+self.nXY[1,0]
        J12 = -self.nXY[0,1]+self.nXY[1,1]
        J21 = -self.nXY[0,0]+self.nXY[2,0]
        J22 = -self.nXY[0,1]+self.nXY[2,1]
        
        self.detJ = J11*J22-J12*J21
        self.invJ = np.array([[J22,-J12],[-J21,J11]])/self.detJ
        
        # Local shape functions
        
        drN = np.array([-1,1,0])
        dsN = np.array([-1,0,1])
        N = lambda r,s: np.array([1-r-s,r,s])
        
        self.dxN = np.repeat([drN*self.invJ[0,0]+dsN*self.invJ[0,1]],4,axis=0).T
        self.dyN = np.repeat([drN*self.invJ[1,0]+dsN*self.invJ[1,1]],4,axis=0).T
        self.N = N(*self.gRS.T)
        
        # Mass and stifness matrix
        
        self.M = np.dot(self.wei*self.detJ*self.N,self.N.T)
        self.Sx = np.dot(self.wei*self.detJ*self.dxN,self.N.T)
        self.Sy = np.dot(self.wei*self.detJ*self.dyN,self.N.T) 

# %% Quadrangles

class Quadrangle:

    def __init__(self,nXY):
        
        self.nXY = nXY
        self.gRS = np.zeros((9,2))
        self.wei = np.array([25,25,25,25,40,40,40,40,64])/81
        
        # Integration points coordinates
        
        self.gRS[0] = [-np.sqrt(3/5),-np.sqrt(3/5)]
        self.gRS[1] = [np.sqrt(3/5),-np.sqrt(3/5)]
        self.gRS[2] = [-np.sqrt(3/5),np.sqrt(3/5)]
        self.gRS[3] = [np.sqrt(3/5),np.sqrt(3/5)]
        self.gRS[4] = [0,-np.sqrt(3/5)]
        self.gRS[5] = [-np.sqrt(3/5),0]
        self.gRS[6] = [np.sqrt(3/5),0]
        self.gRS[7] = [0,np.sqrt(3/5)]
        self.gRS[8] = [0,0]
        
        # Computes the Jacobian matrix
        
        J11 = lambda s: ((s-1)*self.nXY[0,0]+(1-s)*self.nXY[1,0]+(s+1)*self.nXY[2,0]-(s+1)*self.nXY[3,0])/4
        J12 = lambda s: ((s-1)*self.nXY[0,1]+(1-s)*self.nXY[1,1]+(s+1)*self.nXY[2,1]-(s+1)*self.nXY[3,1])/4
        J21 = lambda r: ((r-1)*self.nXY[0,0]-(r+1)*self.nXY[1,0]+(r+1)*self.nXY[2,0]+(1-r)*self.nXY[3,0])/4
        J22 = lambda r: ((r-1)*self.nXY[0,1]-(r+1)*self.nXY[1,1]+(r+1)*self.nXY[2,1]+(1-r)*self.nXY[3,1])/4
        detJ = lambda r,s: J11(s)*J22(r)-J12(s)*J21(r)
        
        invJ11 = lambda r,s: J22(r)/detJ(r,s)
        invJ12 = lambda r,s: -J12(r)/detJ(r,s)
        invJ21 = lambda r,s: -J21(r)/detJ(r,s)
        invJ22 = lambda r,s: J11(r)/detJ(r,s)
        
        # Local shape functions
        
        N = lambda r,s: np.array([(1-r)*(1-s)/4,(1+r)*(1-s)/4,(1+r)*(1+s)/4,(1-r)*(1+s)/4])
        drN = lambda s: np.array([(s-1)/4,(1-s)/4,(s+1)/4,-(s+1)/4])
        dsN = lambda r: np.array([(r-1)/4,-(r+1)/4,(r+1)/4,(1-r)/4])
        dxN = lambda r,s: drN(s)*invJ11(r,s)+dsN(r)*invJ12(r,s)
        dyN = lambda r,s: drN(s)*invJ21(r,s)+dsN(r)*invJ22(r,s)
        
        self.detJ = detJ(*self.gRS.T)
        self.dxN = dxN(*self.gRS.T)
        self.dyN = dyN(*self.gRS.T)
        self.N = N(*self.gRS.T)
        
        # Mass and stifness matrix
        
        self.M = np.dot(self.wei*self.detJ*self.N,self.N.T)
        self.Sx = np.dot(self.wei*self.detJ*self.dxN,self.N.T)
        self.Sy = np.dot(self.wei*self.detJ*self.dyN,self.N.T)  
        
# %% Faces

class Face:
    
    def __init__(self,nXY):
        
        self.M = 0
        self.nXY = nXY
        self.wei = np.array([5/9,5/9,8/9])
        self.gRS = np.array([-np.sqrt(3/5),np.sqrt(3/5),0])
        
        # Outer normal
        
        v = nXY[1]-nXY[0]
        self.norm = np.array([v[1],v[0]])
        self.detJ = np.sqrt(np.sum(v**2))/2
        self.norm /= np.sqrt(np.sum(self.norm**2))
        
        # Mass and local shape functions
        
        self.N = lambda r: np.array([(1-r)/2,(1+r)/2])
        M = lambda r: np.tensordot(self.N(r),self.N(r),axes=0)*self.detJ
        for i in range(3): self.M += self.wei[i]*M(self.gRS[i])
        
    def flux(self,fx,fy):
        
        F = np.abs(self.norm[0]*fx)+np.abs(self.norm[1]*fy)
        F = self.M.dot(F)
        return F

# %% Global Mesh

class Mesh:

    def __init__(self,nXY,eId,fId):
        
        self.nXY = np.array(nXY)
        self.eId = np.array(eId)
        self.fId = np.array(fId)
        
        self.nNbr = self.nXY.shape[0]
        self.eNbr = self.eId.shape[0]
        self.fNbr = self.fId.shape[0]
        
        self.M = sparse.dok_matrix((self.nNbr,self.nNbr))
        self.Sx = sparse.dok_matrix((self.nNbr,self.nNbr))
        self.Sy = sparse.dok_matrix((self.nNbr,self.nNbr))
        self.elems = []
        
        # Creates the boundary faces
        
        self.faces = [Face(self.nXY[self.fId[i]]) for i in range(self.fId.shape[0])]
        self.nfId = np.unique(self.fId.flatten())

        # Creates the domain elements
        
        for i in range(self.eNbr):
            
            if len(self.eId[i])==4:
                
                self.elems.append(Quadrangle(self.nXY[self.eId[i]]))
                
                # Adds to the global mass matrix
                
                self.M[self.eId[i][0],self.eId[i]] += self.elems[i].M[0]
                self.M[self.eId[i][1],self.eId[i]] += self.elems[i].M[1]
                self.M[self.eId[i][2],self.eId[i]] += self.elems[i].M[2]
                self.M[self.eId[i][3],self.eId[i]] += self.elems[i].M[3]
                
                # Adds to the global stifness matrices
        
                self.Sx[self.eId[i][0],self.eId[i]] += self.elems[i].Sx[0]
                self.Sx[self.eId[i][1],self.eId[i]] += self.elems[i].Sx[1]
                self.Sx[self.eId[i][2],self.eId[i]] += self.elems[i].Sx[2]
                self.Sx[self.eId[i][3],self.eId[i]] += self.elems[i].Sx[3]
                
                self.Sy[self.eId[i][0],self.eId[i]] += self.elems[i].Sy[0]
                self.Sy[self.eId[i][1],self.eId[i]] += self.elems[i].Sy[1]
                self.Sy[self.eId[i][2],self.eId[i]] += self.elems[i].Sy[2]
                self.Sy[self.eId[i][3],self.eId[i]] += self.elems[i].Sy[3]
                
            if len(self.eId[i])==3:
                
                self.elems.append(Triangle(self.nXY[self.eId[i]]))
                
                # Adds to the global mass matrix
                
                self.M[self.eId[i][0],self.eId[i]] += self.elems[i].M[0]
                self.M[self.eId[i][1],self.eId[i]] += self.elems[i].M[1]
                self.M[self.eId[i][2],self.eId[i]] += self.elems[i].M[2]
                
                # Adds to the global stifness matrices
        
                self.Sx[self.eId[i][0],self.eId[i]] += self.elems[i].Sx[0]
                self.Sx[self.eId[i][1],self.eId[i]] += self.elems[i].Sx[1]
                self.Sx[self.eId[i][2],self.eId[i]] += self.elems[i].Sx[2]
                
                self.Sy[self.eId[i][0],self.eId[i]] += self.elems[i].Sy[0]
                self.Sy[self.eId[i][1],self.eId[i]] += self.elems[i].Sy[1]
                self.Sy[self.eId[i][2],self.eId[i]] += self.elems[i].Sy[2]
        
        # Converts to optimized sparse
        
        self.Sx = self.Sx.tocsr()
        self.Sy = self.Sy.tocsr()
        self.M = self.M.tocsr()
        
    def flux(self,fx,fy):
        
        F = np.zeros(self.nNbr)
        for i in range(self.fNbr): F[self.fId[i]] += self.faces[i].flux(fx[self.fId[i]],fy[self.fId[i]])
        return F

File: cpp/mesh.cpp
#include <Eigen/Dense>
#include <Eigen/Sparse>
#include <vector>
#include "mesh.h"
#include <omp.h>

using namespace std;
using namespace Eigen;

// Triangle element

Triangle::Triangle(vector<vector<double>> nXY){

    Matrix2d invJ;
    Vector3d drN{-1,1,0};
    Vector3d dsN{-1,0,1};
    Vector4d wei{-27.0/96,25.0/96,25.0/96,25.0/96};
    vector<vector<double>> gRS{{1.0/3,1.0/3},{0.6,0.2},{0.2,0.6},{0.2,0.2}};

    // Computes Jacobian

    double J11 = drN(0)*nXY[0][0]+drN(1)*nXY[1][0]+drN(2)*nXY[2][0];
    double J12 = drN(0)*nXY[0][1]+drN(1)*nXY[1][1]+drN(2)*nXY[2][1];
    double J21 = dsN(0)*nXY[0][0]+dsN(1)*nXY[1][0]+dsN(2)*nXY[2][0];
    double J22 = dsN(0)*nXY[0][1]+dsN(1)*nXY[1][1]+dsN(2)*nXY[2][1];

    double detJ = J11*J22-J12*J21;

    invJ(0,0) = J22/detJ;
    invJ(1,1) = J11/detJ;
    invJ(0,1) = -J12/detJ;
    invJ(1,0) = -J21/detJ;

    N = MatrixXd(3,4);
    MatrixXd dxN(3,4);
    MatrixXd dyN(3,4);

    // Mass and stifness matrices

    for(int i=0; i<4; i++){
        
        N.col(i) = Vector3d {1-gRS[i][0]-gRS[i][1],gRS[i][0],gRS[i][1]};
        dxN.col(i) = drN*invJ(0,0)+dsN*invJ(0,1);
        dyN.col(i) = drN*invJ(1,0)+dsN*invJ(1,1);
    }

    M = N*wei.asDiagonal()*(N.transpose())*detJ;
    Sx = N*wei.asDiagonal()*(dxN.transpose())*detJ;
    Sy = N*wei.asDiagonal()*(dyN.transpose())*detJ;
}

// Quadrangle element

Quadrangle::Quadrangle(vector<vector<double>> nXY){

    VectorXd wei(9);
    vector<vector<double>> gRS;
    
    wei << 25.0/81,25.0/81,25.0/81,25.0/81,40.0/81,40.0/81,40.0/81,40.0/81,64.0/81;
    gRS.push_back({-sqrt(3.0/5),-sqrt(3.0/5)});
    gRS.push_back({sqrt(3.0/5),-sqrt(3.0/5)});
    gRS.push_back({-sqrt(3.0/5),sqrt(3.0/5)});
    gRS.push_back({sqrt(3.0/5),sqrt(3.0/5)});
    gRS.push_back({0,-sqrt(3.0/5)});
    gRS.push_back({-sqrt(3.0/5),0});
    gRS.push_back({sqrt(3.0/5),0});
    gRS.push_back({0,sqrt(3.0/5)});
    gRS.push_back({0,0});

    N = MatrixXd(4,9);
    MatrixXd dxN(4,9);
    MatrixXd dyN(4,9);
    MatrixXd drN(4,9);
    MatrixXd dsN(4,9);
    VectorXd invJ11(9);
    VectorXd invJ12(9);
    VectorXd invJ21(9);
    VectorXd invJ22(9);
    VectorXd detJ(9);

    dxN.setZero();
    dyN.setZero();

    // Computes Jacobian

    auto J11 = [nXY](double s){return ((s-1)*nXY[0][0]+(1-s)*nXY[1][0]+(s+1)*nXY[2][0]-(s+1)*nXY[3][0])/4;};
    auto J12 = [nXY](double s){return ((s-1)*nXY[0][1]+(1-s)*nXY[1][1]+(s+1)*nXY[2][1]-(s+1)*nXY[3][1])/4;};
    auto J21 = [nXY](double r){return ((r-1)*nXY[0][0]-(r+1)*nXY[1][0]+(r+1)*nXY[2][0]+(1-r)*nXY[3][0])/4;};
    auto J22 = [nXY](double r){return ((r-1)*nXY[0][1]-(r+1)*nXY[1][1]+(r+1)*nXY[2][1]+(1-r)*nXY[3][1])/4;};

    for (int i=0; i<9; i++){

        detJ(i) = J11(gRS[i][1])*J22(gRS[i][0])-J12(gRS[i][1])*J21(gRS[i][0]);

        invJ11(i) = J22(gRS[i][0])/detJ(i);
        invJ12(i) = -J12(gRS[i][1])/detJ(i);
        invJ21(i) = -J21(gRS[i][0])/detJ(i);
        invJ22(i) = J11(gRS[i][1])/detJ(i);

        drN.col(i) = Vector4d {(gRS[i][1]-1)/4,(1-gRS[i][1])/4,(gRS[i][1]+1)/4,-(gRS[i][1]+1)/4};
        dsN.col(i) = Vector4d {(gRS[i][0]-1)/4,-(gRS[i][0]+1)/4,(gRS[i][0]+1)/4,(1-gRS[i][0])/4};
    }

    // Local shape functions

    for(int i=0; i<9; i++){

        N(0,i) = (1-gRS[i][0])*(1-gRS[i][1])/4;
        N(1,i) = (1+gRS[i][0])*(1-gRS[i][1])/4;
        N(2,i) = (1+gRS[i][0])*(1+gRS[i][1])/4;
        N(3,i) = (1-gRS[i][0])*(1+gRS[i][1])/4;

        for(int j=0; j<4; j++){
            dxN(j,i) += drN(j,i)*invJ11(i)+dsN(j,i)*invJ12(i);
            dyN(j,i) += drN(j,i)*invJ12(i)+dsN(j,i)*invJ22(i);
        }
    }

    M = N*(wei.asDiagonal()*detJ).asDiagonal()*(N.transpose());
    Sx = N*(wei.asDiagonal()*detJ).asDiagonal()*(dxN.transpose());
    Sy = N*(wei.asDiagonal()*detJ).asDiagonal()*(dyN.transpose());
}

// Face element

Face::Face(vector<vector<double>> nXY,vector<int> fId_in){

    MatrixXd N(2,3);
    Vector3d wei{5.0/9,5.0/9,8.0/9};
    vector<double> gRS{-sqrt(3.0/5),sqrt(3.0/5),0};
    Vector2d v{nXY[1][0]-nXY[0][0],nXY[1][1]-nXY[0][1]};

    // Outer normal
    
    double detJ = sqrt(v(0)*v(0)+v(1)*v(1))/2;
    norm = {v(1)/(2*detJ),v(0)/(2*detJ)};

    // Other attributes

    for(int i=0; i<3; i++){N.col(i) = Vector2d {(1-gRS[i])/2,(1+gRS[i])/2};}
    M = N*wei.asDiagonal()*(N.transpose())*detJ;
    Nf = N*wei*detJ;
    fId = fId_in;
}

Mesh::Mesh(vector<vector<double>> nXY_in,vector<vector<int>> eId_in){

    nXY = nXY_in;
    eId = eId_in;
    nNbr = nXY.size();
    eNbr = eId.size();

    M = SparseMatrix<double>(nNbr,nNbr);
    Sx = SparseMatrix<double>(nNbr,nNbr);
    Sy = SparseMatrix<double>(nNbr,nNbr);

    typedef Triplet<double> T;
    vector<T> tripM;
    vector<T> tripSx;
    vector<T> tripSy;

    #pragma omp parallel
    {
        vector<T> tripMP;
        vector<T> tripSxP;
        vector<T> tripSyP;

        #pragma omp for
        for(int i=0; i<eNbr; i++){
            
            if(eId[i].size()==4){
                Quadrangle elem({nXY[eId[i][0]],nXY[eId[i][1]],nXY[eId[i][2]],nXY[eId[i][3]]});

                for(int j=0; j<4; j++){
                    for(int k=0; k<4; k++){

                        tripMP.push_back(T(eId[i][k],eId[i][j],elem.M(k,j)));
                        tripSxP.push_back(T(eId[i][k],eId[i][j],elem.Sx(k,j)));
                        tripSyP.push_back(T(eId[i][k],eId[i][j],elem.Sy(k,j)));
                    }
                }
            }

            if(eId[i].size()==3){
                Triangle elem({nXY[eId[i][0]],nXY[eId[i][1]],nXY[eId[i][2]]});

                for(int j=0; j<3; j++){
                    for(int k=0; k<3; k++){

                        tripMP.push_back(T(eId[i][k],eId[i][j],elem.M(k,j)));
                        tripSxP.push_back(T(eId[i][k],eId[i][j],elem.Sx(k,j)));
                        tripSyP.push_back(T(eId[i][k],eId[i][j],elem.Sy(k,j)));
                    }
                }
            }
        }

        #pragma omp critical
        {
        tripM.insert(tripM.end(),tripMP.begin(),tripMP.end());
        tripSx.insert(tripSx.end(),tripSxP.begin(),tripSxP.end());
        tripSy.insert(tripSy.end(),tripSyP.begin(),tripSyP.end());
        }
    }
    M.setFromTriplets(tripM.begin(),tripM.end());
    Sx.setFromTriplets(tripSx.begin(),tripSx.end());
    Sy.setFromTriplets(tripSy.begin(),tripSy.end());
}

// Precompute Neumann BC

vector<Face> Mesh::precompute(vector<vector<int>> fId){

    vector<Face> face;
    #pragma omp parallel
    {
        vector<Face> faceP;
        #pragma omp for

        for(int i=0; i<fId.size(); i++){
            Face L2({nXY[fId[i][0]],nXY[fId[i][1]]},fId[i]);
            faceP.push_back(L2);
        }
        #pragma omp critical
        face.insert(face.end(),faceP.begin(),faceP.end());
    }
    return face;
}

// Apply constant Neumann BC

VectorXd Mesh::neumannFix(vector<Face> face, double bc){

    VectorXd B(nNbr);B.setZero();
    for(int i=0; i<face.size(); i++){B(face[i].fId) += face[i].Nf*bc;}
    return B;
}

// Apply variable Neumann BC

VectorXd Mesh::neumannVar(vector<Face> face, vector<VectorXd> F){
    
    VectorXd B(nNbr);
    B.setZero();

    for(int i=0; i<face.size(); i++){
        
        Matrix2d flux;
        flux.col(0) = Vector2d {F[0](face[i].fId[0]),F[0](face[i].fId[1])};
        flux.col(1) = Vector2d {F[1](face[i].fId[0]),F[1](face[i].fId[1])};
        Vector2d bc = face[i].M*flux.cwiseAbs()*face[i].norm.cwiseAbs();
        B(face[i].fId[0]) += bc[0];
        B(face[i].fId[1]) += bc[1];
    }
    return B;
}